lib/utils/set.go

```diff
@@ -0,0 +1,105 @@
+// Teleport
+// Copyright (C) 2024 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package utils
+
+import "maps"
+
+// Set models a collection of unique elements. Its implemented as the classic
+// Go `map[T]struct{}` pattern and has the same underlying representation. This
+// has the nice property of preserving the map-like reference semantics. You can
+// even iterate over the set with `range`.
+type Set[T comparable] map[T]struct{}
+
+// NewSet constructs a set from an arbitrary collection of elements
+func NewSet[T comparable](elements ...T) Set[T] {
+	s := NewSetWithCapacity[T](len(elements))
+	s.AddAll(elements)
+	return s
+}
+
+// NewSetWithCapacity constructs a new, empty set with an given capacity hint.
+func NewSetWithCapacity[T comparable](n int) Set[T] {
+	return make(map[T]struct{}, n)
+}
+
+// Add expands the set to include the supplied element (if not already included),
+// returning a reference to the mutated set.
+func (s Set[T]) Add(element T) Set[T] {
+	s[element] = struct{}{}
+	return s
+}
+
+// AddAll expands the set to include all of the supplied elements.
+func (s Set[T]) AddAll(elements []T) {
+	for _, element := range elements {
+		s[element] = struct{}{}
+	}
+}
+
+// Union updates the set to be the union of the original set and `other`
+func (s Set[T]) Union(other Set[T]) {
```

Comment:
Yeah, I get that. I just didn't want to create a bunch of temporary sets when chaining operations together. Perhaps it's premature optimization.

Reply:
I think that since the `Union` doesn't really return a value, the probability of making a mistake is negligible; after all, you'd have to call `Union` and not care about the result. Perhaps consider adding a separate, receiver-less `Union` function that takes two sets and produces a third one, just for convenience. Up to you.